### PR TITLE
[WIP] #588 - Withdraw Modal

### PR DESCRIPTION
**Pull request by calvinfroedge**
*PLEASE DO NOT MERGE...REVIEW ONLY*

Implements new withdraw modal (#588). Notes:

- Some currencies do not show estimated values (I.e. AGRS). I'm not sure why this is but assuming it's because of limited market data.
- Some currencies have funny behavior (i.e. USDT) when calculating gateway fees

update

### CODE REVIEW

File: app/components/DepositWithdraw/WithdrawModalNew.jsx [hunk 1]
```diff
@@ -0,0 +1,834 @@
...
+
+          {/*WITHDRAW ADDRESS*/}
+          {
+            (assetAndGateway && !isBTS) ? 
+            <div style={{marginBottom: "1em"}}>
+              <label className="left-label">
+                  <Translate component="span" content="modal.withdraw.address"/>
+              </label>
+              {addressError ? <div className="has-error" style={{position: "absolute", right: "1em", marginTop: "-30px"}}>
+                  <Translate content="modal.withdraw.address_not_valid" coin_type={selectedAsset} />
```
Comment (startailcoon) [inline app/components/DepositWithdraw/WithdrawModalNew.jsx]: Missing `coin_type` in translation file 
`Warning: Unknown prop `coin_type` on <span> tag.`

File: app/components/DepositWithdraw/WithdrawModalNew.jsx [hunk 2]
```diff
@@ -0,0 +1,834 @@
...
+              </label>
+              <ExchangeInput value={quantity} onChange={this.onQuantityChanged.bind(this)} onFocus={onFocus} onBlur={onBlur} />
+            </div> : 
+            null
+          }
+
+          {
+            (assetAndGateway || isBTS) && !canCoverWithdrawal ? 
+            <div style={{marginBottom: "1em"}}>
+              <Translate content="modal.withdraw.cannot_cover" />
```
Comment (startailcoon) [inline app/components/DepositWithdraw/WithdrawModalNew.jsx]: `missing translation: en.modal.withdraw.cannot_cover`

File: app/components/DepositWithdraw/WithdrawModalNew.jsx [hunk 3]
```diff
@@ -0,0 +1,834 @@
...
+        }, this._updateFee);
+    }
+
+    onAssetSelected(value, asset){
+        let { selectedAsset, selectedGateway } = _onAssetSelected.call(this, value);
+        let address = WithdrawAddresses.getLast(value.toLowerCase());
+        this.setState({selectedAsset, selectedGateway, gateFee: asset.gateFee, issuer: asset.issuer, address, isBTS: false});
+    }
+
+    onAssetChanged(value){
```
Comment (startailcoon) [inline app/components/DepositWithdraw/WithdrawModalNew.jsx]: Should clear all input fields as we're switching asset. Clear Amount, Address and Memo

File: app/components/DepositWithdraw/DepositWithdrawAssetSelector.js [hunk 4]
```diff
@@ -15,6 +15,7 @@ class DepositWithdrawAssetSelector  extends React.Component {
 
     render(){
         const { props } = this;
+        const { include } = props;
```
Comment (startailcoon) [inline app/components/DepositWithdraw/DepositWithdrawAssetSelector.js]: Change this to 
`const { include, openLedgerBackedCoins, rudexBackedCoins, blockTradesBackedCoins } = this.props`

Then remove the line before and update L53 correspondently.

File: app/lib/common/assetGatewayMixin.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    return gatewayStatus;
+}
+
+function _openGatewaySite() {
+    let {selectedGateway, gatewayStatus} = this.state;
+    let win = window.open(gatewayStatus[selectedGateway].support_url, "_blank");
+    win.focus();
+}
+
+function _getNumberAvailableGateways(){
```
Comment (startailcoon) [inline app/lib/common/assetGatewayMixin.js]: This needs to accommodate if the user has a withdrawable asset or not for the gateway as well.

File: app/components/DepositWithdraw/DepositWithdrawAssetSelector.js (hunk 4 above)
Comment (startailcoon) [reply app/components/DepositWithdraw/DepositWithdrawAssetSelector.js]: I've noticed you've worked a bit different than what I've seen on most other places of the app. First setting props and state. At least add the coins to the props line and update L53 for consistency.

File: app/lib/common/assetGatewayMixin.js (hunk 5 above)
Comment (calvinfroedge) [reply app/lib/common/assetGatewayMixin.js]: You're going to need to help me on that

File: app/components/DepositWithdraw/DepositWithdrawAssetSelector.js (hunk 4 above)
Comment (calvinfroedge) [reply app/components/DepositWithdraw/DepositWithdrawAssetSelector.js]: Good catch

File: app/components/DepositWithdraw/WithdrawModalNew.jsx (hunk 3 above)
Comment (calvinfroedge) [reply app/components/DepositWithdraw/WithdrawModalNew.jsx]: Check

File: app/components/DepositWithdraw/WithdrawModalNew.jsx (hunk 1 above)
Comment (calvinfroedge) [reply app/components/DepositWithdraw/WithdrawModalNew.jsx]: I changed the message and removed this attribute

File: app/components/DepositWithdraw/WithdrawModalNew.jsx (hunk 2 above)
Comment (calvinfroedge) [reply app/components/DepositWithdraw/WithdrawModalNew.jsx]: Must have gotten lost in the merge

**Comment by calvinfroedge** [comment]
User should not be able to change gateway, so I disabled that all together

**Comment by calvinfroedge** [comment]
@startailcoon Changes implemented

**Comment by wmbutler** [comment]
There are a lot of fundamental problems with the withdraw modal. I think you may not fully understand how it works in the context of the DEX and gateways. A withdrawal is actually just a transfer to a gateway provider with a specially coded memo sort of like the way banks handle "with further routing to". For instance, when openledger accepts BTC withdrawal, the funds are actually sent to `openledger-dex` with a memo of `bts:1490r3409ri3409ri340ir3`.

Since you are showing the memo field for withdrawals, you are really setting the user up for failure. Have you considered looking at the logic for the existing withdraw modal? A lot of it is done correctly but we need a more flexible way to incorporate other gateway providers. It even appears to have validation for some currencies.

withdraw_legacy

Other comments:

- [ ] Not many currencies are available, in particular I could not select STEEM. I verified that I had .5 STEEM in my active account.
- [ ] Can't click on the carat in the asset dropdown. This may be a limitation of the new typeahade widget.

**Comment by calvinfroedge** [comment]
Can you tell me, based on the data structures available, how I can
programmatically determine if the memo field should be shown?

**Comment by wmbutler** [comment]
Each gateway has it's own way of handling it, but openledger handles it pretty much the way I showed you in the screen captures above, particularly the modal that currently exists in the bitshares wallet.

**Comment by calvinfroedge** [comment]
There is a modal for each gateway in the old code. I'll check the rules.

**Comment by wmbutler** [comment]
Actually the modal in the old code just does openledger I think.

**Comment by calvinfroedge** [comment]
@wmbutler Memo field is disabled for all OPEN based assets. The reason STEEM was not appearing in the withdraw modal is that it is not a backed coin. It does not appear in the old withdrawal modal either.

**Comment by calvinfroedge** [comment]
update

Added withdraw modal to dashboard / portfolio assets.

**Comment by startailcoon** [comment]
Steem is a backed asset, but I think they all have it disabled for the moment.

**Comment by wmbutler** [comment]
Another precision issue but now for LTC. (1LTC is definitely not worth 1 million BTS). I think you need to know the precision of each asset in order to do this comparison. Should we just eliminate the equivalent field for now since we are so close to release date?

screen shot 2018-02-27 at 3 15 44 pm

**Comment by svk31** [comment]
What's the current status here? There's quite a lot of conflicts, should I try to resolve them in order to test it or can you handle it first?

**Comment by wmbutler** [comment]
@svk31 from a functionality state, it's good. Regarding conflicts, I'm guessing it's because this branch was initially created so long ago. @calvinfroedge can you help with resolution?

**Pull request merged**